src/main/java/org/jboss/jandex/GenericSignatureParser.java

```diff
@@ -406,6 +406,11 @@ private Type parseTypeParameter() {
         // Class bound has an optional reference type
         if (signature.charAt(pos) != ':') {
             bounds.add(parseReferenceType());
+        } else {
+            // Pretend optional reference type is Object
+            // See https://github.com/openjdk/jdk/commit/9d7f8bc9a053595c290ae339b902dc762b848ab9
+            // See https://github.com/wildfly/jandex/issues/99
+            bounds.add(ClassType.OBJECT_TYPE);
```

Comment:
By the way, I don't know if this has implications, however, it seems to be the least evil. Otherwise, jandex index format must be improved to support the distinction between `<T extends Object & Interface>` vs `<T extends Interface>` (which have `T:Object:Interface` and `T::Interface` signatures).

However, I guess inserting `Object` here allows accessing the annotations without updating the index format (e.g. to account for optional yet absent class bound).

My best guess is that optional class bounds are allowed here to allow users to configure the erasure signatures.

For instance `<T extends Number>` erases to `Number` while `<T extends Object & Number>` erases to `Object` which allows Java programmres to evolve old code (e.g. keep backward-compatible API while adding generics to their code).

^^^ The above is a wild guess, so I am not sure if it should be added to the code comment.

Reply:
So I agree it's a good solution: there is something to be said for having consistent  semantic representation vs the multiple encodings. However, since I would like to include this fix as a bug fix release, I was a little worried about behavioral compatibility differences that might impact some users. For example, previous code might have expected T extends Runnable to have .bounds().get(0) return Runnable, whereas after this change it would be Object.  So I changed the approach to instead use a hidden marker flag only for the indexing process, and in that way the implied object class bound is hidden. 

Your wild guess is correct, the JLS does define erasure to be the left most class bound, so Object & Interface is functionally different than just Interface.

I have updated the PR with my changes. I intend to merge tomorrow provided you have no concerns or questions with the update